main/modes/mode_dance.c

```diff
@@ -0,0 +1,1247 @@
+    {.func = danceRise,  .arg = RGB_2_ARG(0, 0, 0xFF), .name = "Rise B"},
+    {.func = danceRise,  .arg = RGB_2_ARG(0, 0, 0),    .name = "Rise RGB"},
+    {.func = dancePulse, .arg = RGB_2_ARG(0xFF, 0, 0), .name = "Pulse R"},
+    {.func = dancePulse, .arg = RGB_2_ARG(0, 0xFF, 0), .name = "Pulse G"},
+    {.func = dancePulse, .arg = RGB_2_ARG(0, 0, 0xFF), .name = "Pulse B"},
+    {.func = dancePulse, .arg = RGB_2_ARG(0, 0, 0),    .name = "Pulse RGB"},
+    {.func = danceSharpRainbow,  .arg = 0, .name = "Rainbow Sharp"},
+    {.func = danceSmoothRainbow, .arg = 20000, .name = "Rainbow Slow"},
+    {.func = danceSmoothRainbow, .arg =  4000, .name = "Rainbow Fast"},
+    {.func = danceRainbowSolid,  .arg = 0, .name = "Rainbow Solid"},
+    {.func = danceFire, .arg = RGB_2_ARG(0xFF, 51, 0), .name = "Fire R"},
+    {.func = danceFire, .arg = RGB_2_ARG(0, 0xFF, 51), .name = "Fire G"},
+    {.func = danceFire, .arg = RGB_2_ARG(51, 0, 0xFF), .name = "Fire B"},
+    {.func = danceBinaryCounter, .arg = 0, .name = "Binary"},
+    {.func = dancePoliceSiren,   .arg = 0, .name = "Siren"},
+    {.func = dancePureRandom,    .arg = 0, .name = "Random"},
+    {.func = danceChristmas,     .arg = 1, .name = "Holiday 1"},
+    {.func = danceChristmas,     .arg = 0, .name = "Holiday 2"},
+    {.func = danceNone,          .arg = 0, .name = "None"},
+    {.func = danceRandomDance,   .arg = 0, .name = "???"},
+};
+
+typedef struct
+{
+    display_t* disp;
+
+    uint8_t danceIdx;
+
+    bool updateScreen;
+    bool resetDance;
+    bool blankScreen;
+
+    font_t infoFont;
+} danceMode_t;
+
+
+swadgeMode modeDance =
+{
+    .modeName = "Light Dances",
+    .fnEnterMode = danceEnterMode,
+    .fnExitMode = danceExitMode,
+    .fnMainLoop = danceMainLoop,
+    .fnButtonCallback = danceButtonCb,
+    .fnTouchCallback = NULL,
+    .wifiMode = NO_WIFI,
+    .fnEspNowRecvCb = NULL,
+    .fnEspNowSendCb = NULL,
+    .fnAccelerometerCallback = NULL,
+    .fnAudioCallback = NULL,
+    .fnTemperatureCallback = NULL,
+};
+
+danceMode_t* danceState;
+
+/*============================================================================
+ * Functions
+ *==========================================================================*/
+
+void danceEnterMode(display_t* disp)
+{
+    danceState = calloc(1, sizeof(danceMode_t));
+
+    danceState->disp = disp;
+
+    danceState->danceIdx = 0;
+
+    danceState->updateScreen = true;
+    danceState->resetDance = true;
+    danceState->blankScreen = false;
+
+    // TODO move font to var
+    if (!loadFont("radiostars.font", &(danceState->infoFont)))
+    {
+        ESP_LOGE("Dance", "Error loading radiostars.font");
+    }
+}
+
+void danceExitMode()
+{
+    freeFont(&(danceState->infoFont));
+    free(danceState);
+    danceState = NULL;
+}
+
+void danceMainLoop(int64_t elapsedUs)
+{
+    ledDances[danceState->danceIdx].func(elapsedUs, ledDances[danceState->danceIdx].arg, danceState->resetDance);
+
+    if (danceState->updateScreen)
```

Comment:
I'd get rid of this and just draw each loop. If you hold start+select, the system draws a progress bar to exit the mode, and if the mode's screen isn't redrawn with reasonable periodicity, it doesn't work well.